File: tools/analytics.mjs
// Prove what the Google Analytics loader does, in a browser-shaped DOM.
//
// `homeroom.analytics` adds GA4 to the published pages after rendering (owner
// decision, 2026-09-17). `make analytics-gate` runs it over a copy of the fixture
// build, and this reads that copy: its `analytics.js`, and one page of each shape
// (the landing page, a school page in each language, a county page, an ask page).
//
// For each scenario the loader is run in jsdom against a stubbed navigator,
// location and storage, and the result is read back:
//
//   * served from homeroom.chelseakr.com, no signal: exactly one gtag.js script,
//     the Consent Mode v2 defaults and the config in order, one page view whose
//     address is the path plus utm_* only;
//   * Global Privacy Control, each Do Not Track form, the stored opt-out, and any
//     other host: no dataLayer, no script, no kill switch -- nothing;
//   * the footer button: opts out (flag stored, cookies removed, label and
//     status change), opts back in (GA starts), in both languages;
//   * axe finds nothing on each page with the button rendered.
//
// Then the negative controls: the GPC guard and the host guard are removed from a
// copy of the loader, the removal is asserted to have happened exactly once, and
// the same scenario must now see GA load. A harness that cannot see a missing
// guard would pass a loader without one, so it has to prove it can.
//
// Usage: node tools/analytics.mjs <site-directory-with-analytics-added>

import { readFileSync, readdirSync, existsSync } from "node:fs";
import { join } from "node:path";
import { JSDOM, VirtualConsole } from "jsdom";
import axe from "axe-core";

const site = process.argv[2];
if (!site) {
  console.error("usage: node tools/analytics.mjs <site-directory>");
  process.exit(2);
}
const loaderPath = join(site, "analytics.js");
if (!existsSync(loaderPath)) {
  console.error(`no ${loaderPath}; run python -m homeroom.analytics over the tree first`);
  process.exit(2);
}
const LOADER = readFileSync(loaderPath, "utf8");
const ID = LOADER.match(/var ID = "(G-[A-Z0-9]+)";/)?.[1];
if (!ID) {
  console.error("analytics.js names no measurement ID");
  process.exit(2);
}
const HOST = "homeroom.chelseakr.com";

function firstPage(dir, suffix) {
  const found = readdirSync(join(site, dir))
    .filter((name) => name.endsWith(suffix))
    .sort()[0];
  if (!found) throw new Error(`no ${suffix} page in ${dir || "."}`);
  return join(dir, found);
}
const PAGES = {
  landing: "index.html",
  schoolEn: firstPage("", ".en.html"),
  schoolEs: firstPage("", ".es.html"),
  county: firstPage("county", ".en.html"),
  ask: firstPage("ask", ".es.html"),
};

let failures = 0;
function check(name, ok, detail = "") {
  if (!ok) failures += 1;
  console.log(`${ok ? "pass" : "FAIL"}  ${name}${ok || !detail ? "" : ` -- ${detail}`}`);
}

/** One page load with the loader run in it. */
function visit(page, { host = HOST, search = "", gpc, dnt, dntWhere = "navigator", stored, blockStorage, loader = LOADER } = {}) {
  const html = readFileSync(join(site, page), "utf8");
  const console_ = new VirtualConsole();
  console_.on("jsdomError", (error) => {
    if (!/getContext\(\) method/.test(error.message)) console.error(error.message);
  });
  const dom = new JSDOM(html, {
    url: `https://${host}/${page}${search}`,
    runScripts: "outside-only",
    pretendToBeVisual: true,
    virtualConsole: console_,
  });
  const { window } = dom;
  if (gpc !== undefined) Object.defineProperty(window.navigator, "globalPrivacyControl", { value: gpc });
  if (dnt !== undefined) {
    const target = dntWhere === "window" ? window : window.navigator;
    Object.defineProperty(target, dntWhere === "ms" ? "msDoNotTrack" : "doNotTrack", { value: dnt });
  }
  if (stored) window.localStorage.setItem("homeroom.chelseakr.com:analytics-opt-out", "1");
  if (blockStorage) {
    Object.defineProperty(window, "localStorage", {
      get() {
        throw new window.DOMException("blocked", "SecurityError");
      },
    });
  }
  window.document.cookie = "_ga=GA1.1.1.1; path=/";
  window.document.cookie = `_ga_${ID.slice(2)}=GS1.1.1; path=/`;
  window.eval(loader);
  return dom;
}

function layer(dom) {
  return (dom.window.dataLayer ?? []).map((entry) => JSON.parse(JSON.stringify(Array.from(entry))));
}
function gtagScripts(dom) {
  return Array.from(dom.window.document.querySelectorAll('script[src*="googletagmanager.com"]'));
}
function nothingLoaded(dom) {
  return (
    dom.window.dataLayer === undefined &&
    gtagScripts(dom).length === 0 &&
    Object.getOwnPropertyDescriptor(dom.window, `ga-disable-${ID}`) === undefined
  );
}

// ---- Loads, with the agreed configuration ------------------------------------

{
  const dom = visit(PAGES.schoolEn, { search: "?utm_source=flyer&student=Ada" });
  const commands = layer(dom);
  const scripts = gtagScripts(dom);
  check("production host, no signal: one async gtag.js for the ID", scripts.length === 1 && scripts[0].async && scripts[0].src === `https://www.googletagmanager.com/gtag/js?id=${ID}`);
  const [regional, global, redaction, js, config, set, event] = commands;
  check("consent: EEA/UK/CH analytics denied, ads denied", regional?.[0] === "consent" && regional[2].analytics_storage === "denied" && regional[2].ad_user_data === "denied" && regional[2].region.includes("DE") && regional[2].region.includes("GB") && regional[2].region.includes("CH") && !regional[2].region.includes("US"));
  check("consent: analytics granted elsewhere, ads still denied", global?.[2].analytics_storage === "granted" && global[2].ad_storage === "denied" && global[2].ad_personalization === "denied");
  check("ads_data_redaction then js", redaction?.[1] === "ads_data_redaction" && js?.[0] === "js");
  check("config: signals and ad personalization off, manual page view", config?.[0] === "config" && config[1] === ID && config[2].allow_google_signals === false && config[2].allow_ad_personalization_signals === false && config[2].send_page_view === false);
  check("one page view, address = path + utm only", set?.[0] === "set" && set[1].page_location === `https://${HOST}/${PAGES.schoolEn}?utm_source=flyer` && event?.[1] === "page_view" && commands.length === 7, JSON.stringify(set));
  check("the kill switch reads the opt-out live", dom.window[`ga-disable-${ID}`] === false);
  check("cookies left alone when not opted out", dom.window.document.cookie.includes("_ga="));
  dom.window.close();
}

// ---- Loads nothing ------------------------------------------------------------

for (const [label, options] of [
  ["Global Privacy Control", { gpc: true }],
  ["Do Not Track (navigator)", { dnt: "1" }],
  ['Do Not Track "yes"', { dnt: "yes" }],
  ["Do Not Track (window)", { dnt: "1", dntWhere: "window" }],
  ["Do Not Track (msDoNotTrack)", { dnt: "1", dntWhere: "ms" }],
  ["the stored opt-out", { stored: true }],
  ["another host (a fixture or preview)", { host: "homeroom.example" }],
  ["localhost", { host: "localhost" }],
]) {
  const dom = visit(PAGES.schoolEs, options);
  const w = dom.window;
  const landed =
    (options.gpc === undefined || w.navigator.globalPrivacyControl === true) &&
    (options.dnt === undefined || [w.navigator.doNotTrack, w.doNotTrack, w.navigator.msDoNotTrack].includes(options.dnt)) &&
    (!options.stored || w.localStorage.getItem("homeroom.chelseakr.com:analytics-opt-out") === "1") &&
    (!options.host || w.location.hostname === options.host);
  check(`${label}: the setup took effect`, landed);
  check(`${label}: nothing loads`, nothingLoaded(dom), JSON.stringify(layer(dom)).slice(0, 200));
  if (options.stored || options.gpc || options.dnt) {
    check(`${label}: this site's GA cookies removed`, !/(^|; )_ga=/.test(w.document.cookie) && !w.document.cookie.includes(`_ga_${ID.slice(2)}=`));
  }
  dom.window.close();
}

// ---- The button -----------------------------------------------------------------

for (const [page, lang, optOut, optIn, off] of [
  [PAGES.schoolEn, "en", "Opt out of analytics", "Opt back in", "Analytics is off in this browser."],
  [PAGES.schoolEs, "es", "Desactivar las analíticas", "Volver a activarlas", "Las analíticas están desactivadas en este navegador."],
  [PAGES.county, "en", "Opt out of analytics", "Opt back in", "Analytics is off in this browser."],
  [PAGES.ask, "es", "Desactivar las analíticas", "Volver a activarlas", "Las analíticas están desactivadas en este navegador."],
]) {
  const dom = visit(page);
  const d = dom.window.document;
  const buttons = d.querySelectorAll(".analytics-opt-out button");
  check(`${page}: one button, labelled in ${lang}`, buttons.length === 1 && buttons[0].textContent === optOut, buttons[0]?.textContent);
  buttons[0].click();
  check(`${page}: opting out stores the flag and relabels`, dom.window.localStorage.getItem("homeroom.chelseakr.com:analytics-opt-out") === "1" && buttons[0].textContent === optIn && d.querySelector(".analytics-opt-out [role=status]").textContent === off);
  check(`${page}: opting out removes the GA cookies and stops hits`, !/(^|; )_ga=/.test(d.cookie) && dom.window[`ga-disable-${ID}`] === true);
  buttons[0].click();
  check(`${page}: opting back in clears the flag`, dom.window.localStorage.getItem("homeroom.chelseakr.com:analytics-opt-out") === null && buttons[0].textContent === optOut);
  dom.window.close();
}

{
  const dom = visit(PAGES.schoolEn, { stored: true });
  const button = dom.window.document.querySelector(".analytics-opt-out button");
  check("a later page reads the stored choice", button?.textContent === "Opt back in" && nothingLoaded(dom));
  button.click();
  check("opting back in on a page where GA never loaded starts it", gtagScripts(dom).length === 1 && layer(dom).some((c) => c[0] === "event" && c[1] === "page_view"));
  dom.window.close();
}

{
  const dom = visit(PAGES.landing);
  const buttons = dom.window.document.querySelectorAll(".analytics-opt-out button");
  const sections = Array.from(buttons).map((b) => b.closest("[lang]").getAttribute("lang"));
  check("landing: the disclosure carries a button in each language", buttons.length === 2 && sections.join() === "en,es" && buttons[1].textContent === "Desactivar las analíticas");
  check("landing: both disclosures are there", Boolean(dom.window.document.getElementById("privacy-en")) && Boolean(dom.window.document.getElementById("privacy-es")));
  dom.window.close();
}

{
  const dom = visit(PAGES.schoolEn, { blockStorage: true });
  check("storage blocked: no button, since the choice could not be kept", dom.window.document.querySelectorAll(".analytics-opt-out button").length === 0 && dom.window.document.querySelector(".analytics-opt-out").hidden === true);
  dom.window.close();
}

// ---- Accessibility of what the loader adds ---------------------------------------

for (const page of Object.values(PAGES)) {
  const dom = visit(page);
  dom.window.eval(axe.source);
  const results = await dom.window.axe.run(dom.window.document, {
    runOnly: { type: "tag", values: ["wcag2a", "wcag2aa", "wcag21a", "wcag21aa", "wcag22aa", "best-practice"] },
    resultTypes: ["violations"],
  });
  const violations = results.violations.filter((v) => v.id !== "color-contrast");
  check(`${page}: axe finds nothing with the button rendered`, violations.length === 0, violations.map((v) => v.id).join(", "));
  dom.window.close();
}

// ---- Negative controls: the harness must see a missing guard ---------------------

for (const [label, guard, options] of [
  ["GPC guard", "    if (n.globalPrivacyControl === true) return true;\n", { gpc: true }],
  ["host guard", "|| w.location.hostname !== HOST ", { host: "homeroom.example" }],
]) {
  const count = LOADER.split(guard).length - 1;
  check(`negative control (${label}): the guard is in the loader exactly once`, count === 1, `found ${count}`);
  const sabotaged = LOADER.replace(guard, "");
  check(`negative control (${label}): the removal landed`, sabotaged !== LOADER && !sabotaged.includes(guard));
  const dom = visit(PAGES.schoolEn, { ...options, loader: sabotaged });
  check(`negative control (${label}): without it, the harness sees GA load`, !nothingLoaded(dom));
  dom.window.close();
}

if (failures) {
  console.error(`analytics: ${failures} check(s) failed`);
  process.exit(1);
}
console.log("analytics: every check passed");

File: src/homeroom/analytics.py
"""Google Analytics 4 on the published site, added after rendering.

Owner decision, 2026-09-17: GA4 on every public site, with the privacy copy and
every "no tracking" claim updated to match. This reverses the rule that the
school pages carry no script (AGENTS.md rule 7, ADR 0001), on the owner's
instruction, and the documents that stated it now say what runs.

It is a step between rendering and serving, not a change to the renderer. The
renderer still writes pages with no script, and every gate that reads its output
(`make pages` over the fixture build, `tools/ask-optin.mjs`'s zero-requests
check) still reads exactly that. `make publish` then runs :func:`add_to_tree`
over its staging tree before weighing it, so the committed ``site/`` -- the
bytes served -- carries the addition, and `tests/test_published_site.py` holds
every published page to it.

What it adds, and nothing else:

* ``analytics.js`` at the site root: the loader, with the measurement ID, the
  production host and the opt-out labels written into it.
* One ``<script src="analytics.js" defer integrity="sha384-...">`` in each
  page's ``<head>``: a relative path, like every other link on the site, and
  pinned to the loader's own hash, so a browser refuses an ``analytics.js``
  that is not the one the page was published with.
* On every page but the landing page, one short note -- "This site uses Google
  Analytics..." -- with a link to the disclosure and an empty slot the loader
  fills with the "Opt out of analytics" button. It goes at the end of the
  footer where a page has one, and at the end of ``<main>`` where the page
  closes on its notices instead (county and district pages).
* On the landing page, the full disclosure in each language section
  (``#privacy-en``, ``#privacy-es``), with the button.

The loader itself refuses to do anything unless the page is served from
``homeroom.chelseakr.com``, the browser sends neither Global Privacy Control nor
Do Not Track, and the reader has not opted out. It sets Consent Mode v2
defaults (ad storage, ad user data and ad personalization denied everywhere;
analytics storage denied in the EEA, the UK and Switzerland), turns Google
signals and ad personalization off, and sends one page view whose address is
the path plus any ``utm_*`` tags and nothing else. The site is not a
single-page app: every page is a document load.

With :data:`GA4_MEASUREMENT_ID` empty, :func:`add_to_tree` changes nothing and
writes nothing, so turning GA off is one line and a ``make publish``.
"""

from __future__ import annotations

import argparse
import base64
import hashlib
import json
import re
from pathlib import Path

from homeroom.i18n import LOCALES, Locale, text

GA4_MEASUREMENT_ID = "G-PMC113MW2C"
"""The web stream of GA4 property 554882251 (14-month retention, Google signals
off on the property). Public by design: it is in every page view's request."""

PRODUCTION_HOST = "homeroom.chelseakr.com"
"""The only host the loader runs on. A local preview, a fixture build and any
other copy of these files load nothing."""

OPT_OUT_KEY = "homeroom.chelseakr.com:analytics-opt-out"
"""Where the footer choice is remembered, in the reader's localStorage."""

SCRIPT_NAME = "analytics.js"
MARKER = 'data-analytics="ga4"'
"""On everything this module adds, so a second run can tell it already ran."""

_MEASUREMENT_ID = re.compile(r"^G-[A-Z0-9]{4,20}$")

CONSENT_DENIED_REGIONS: tuple[str, ...] = (
    # EU member states
    "AT", "BE", "BG", "HR", "CY", "CZ", "DK", "EE", "FI", "FR", "DE", "GR", "HU",
    "IE", "IT", "LV", "LT", "LU", "MT", "NL", "PL", "PT", "RO", "SK", "SI", "ES",
    "SE",
    # EU territories with their own ISO codes
    "AX", "GF", "GP", "MQ", "MF", "RE", "YT",
    # Rest of the EEA, the UK and Switzerland
    "IS", "LI", "NO", "GB", "CH",
)  # fmt: skip
"""Where Consent Mode denies ``analytics_storage`` by default (ISO 3166-1)."""

_HEAD_CLOSE = "</head>"
_FOOTER_END = "</div>\n</footer>"
_MAIN_END = "</main>"
_LANDING_SECTION_END = "</section>"
_HTML_LANG = re.compile(r'<html lang="(en|es)">')

LOADER = """/* Google Analytics 4 for homeroom.chelseakr.com. Added to the published pages by
   homeroom.analytics (owner decision, 2026-09-17). Loads nothing off the production
   host, under Global Privacy Control or Do Not Track, or after "Opt out of
   analytics". */
(function () {
  "use strict";
  var ID = __ID__;
  var HOST = __HOST__;
  var KEY = __KEY__;
  var REGIONS = __REGIONS__;
  var LABELS = __LABELS__;
  var w = window;
  var d = document;
  var n = navigator;

  function signal() {
    if (n.globalPrivacyControl === true) return true;
    var dnt = n.doNotTrack || w.doNotTrack || n.msDoNotTrack;
    return dnt === "1" || dnt === "yes";
  }
  function stored() {
    try {
      return w.localStorage.getItem(KEY) === "1";
    } catch (e) {
      return false;
    }
  }
  function optedOut() {
    return signal() || stored();
  }
  function scrubbedPath(path) {
    return path.split("/").map(function (part) {
      return part === "" || (part.length <= 100 && /^[a-z0-9]+(?:[-_.][a-z0-9]+)*$/i.test(part))
        ? part : ":redacted";
    }).join("/");
  }
  function campaign(search) {
    var kept = [];
    var keys = ["utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content", "utm_id"];
    var params = new URLSearchParams(search);
    for (var i = 0; i < keys.length; i++) {
      var value = params.get(keys[i]);
      if (value) kept.push(keys[i] + "=" + encodeURIComponent(value.slice(0, 100)));
    }
    return kept.length ? "?" + kept.join("&") : "";
  }
  function referrer() {
    if (!d.referrer) return "";
    try {
      var url = new URL(d.referrer);
      if (url.origin === w.location.origin) return url.origin + scrubbedPath(url.pathname) + campaign(url.search);
      return url.origin + "/";
    } catch (e) {
      return "";
    }
  }
  function clearCookies() {
    var own = { _ga: true };
    own["_ga_" + ID.slice(2)] = true;
    var labels = w.location.hostname.split(".");
    var domains = [""];
    for (var i = 0; i < labels.length - 1; i++) domains.push(labels.slice(i).join("."));
    d.cookie.split(";").forEach(function (pair) {
      var name = pair.split("=")[0].trim();
      if (!own[name]) return;
      domains.forEach(function (domain) {
        d.cookie = name + "=; Max-Age=0; path=/" + (domain ? "; domain=" + domain : "");
      });
    });
  }

  var loaded = false;
  function load() {
    if (loaded || !ID || w.location.hostname !== HOST || optedOut()) return false;
    loaded = true;
    Object.defineProperty(w, "ga-disable-" + ID, { configurable: true, get: optedOut });
    var layer = (w.dataLayer = w.dataLayer || []);
    function gtag() {
      layer.push(arguments);
    }
    gtag("consent", "default", {
      ad_storage: "denied", ad_user_data: "denied", ad_personalization: "denied",
      analytics_storage: "denied", region: REGIONS
    });
    gtag("consent", "default", {
      ad_storage: "denied", ad_user_data: "denied", ad_personalization: "denied",
      analytics_storage: "granted"
    });
    gtag("set", "ads_data_redaction", true);
    gtag("js", new Date());
    gtag("config", ID, {
      send_page_view: false, allow_google_signals: false, allow_ad_personalization_signals: false
    });
    var page = {
      page_location: w.location.origin + scrubbedPath(w.location.pathname) + campaign(w.location.search),
      page_title: d.title.slice(0, 300)
    };
    var from = referrer();
    if (from) page.page_referrer = from;
    gtag("set", page);
    gtag("event", "page_view");
    var script = d.createElement("script");
    script.async = true;
    script.src = "https://www.googletagmanager.com/gtag/js?id=" + encodeURIComponent(ID);
    d.head.appendChild(script);
    return true;
  }

  function storageWorks() {
    try {
      var current = w.localStorage.getItem(KEY);
      w.localStorage.setItem(KEY, current === null ? "0" : current);
      if (current === null) w.localStorage.removeItem(KEY);
      return true;
    } catch (e) {
      return false;
    }
  }
  function renderControls(changed) {
    var slots = d.querySelectorAll(".analytics-opt-out");
    var off = stored();
    for (var i = 0; i < slots.length; i++) {
      var slot = slots[i];
      var holder = slot.closest("[lang]");
      var labels = LABELS[holder && LABELS[holder.getAttribute("lang")] ? holder.getAttribute("lang") : "en"];
      var button = slot.querySelector("button");
      var status = slot.querySelector("[role=status]");
      if (!button) {
        button = d.createElement("button");
        button.type = "button";
        button.style.font = "inherit";
        button.style.minHeight = "24px";
        button.style.marginInlineStart = "0.5em";
        button.style.cursor = "pointer";
        button.addEventListener("click", toggle);
        status = d.createElement("span");
        status.setAttribute("role", "status");
        status.style.marginInlineStart = "0.5em";
        slot.appendChild(button);
        slot.appendChild(status);
      }
      button.textContent = off ? labels.optIn : labels.optOut;
      status.textContent = off ? labels.off : changed ? labels.on : "";
      slot.hidden = false;
    }
  }
  function toggle() {
    var optOut = !stored();
    try {
      if (optOut) w.localStorage.setItem(KEY, "1");
      else w.localStorage.removeItem(KEY);
    } catch (e) {
      return;
    }
    if (optOut) clearCookies();
    else load();
    renderControls(true);
  }

  if (optedOut() && ID) clearCookies();
  load();
  if (storageWorks()) renderControls(false);
})();
"""


def measurement_id(raw: str = GA4_MEASUREMENT_ID) -> str | None:
    """The configured ID, or ``None`` when it is empty (GA off)."""
    value = raw.strip()
    if not value:
        return None
    if not _MEASUREMENT_ID.fullmatch(value):
        raise ValueError(f"not a GA4 measurement ID: {value!r}")
    return value


def loader(measurement: str) -> str:
    """The text of ``analytics.js`` for one measurement ID."""
    labels = {
        locale: {
            "optOut": text(locale, "analytics_opt_out"),
            "optIn": text(locale, "analytics_opt_in"),
            "off": text(locale, "analytics_off_status"),
            "on": text(locale, "analytics_on_status"),
        }
        for locale in LOCALES
    }
    values = {
        "__ID__": json.dumps(measurement),
        "__HOST__": json.dumps(PRODUCTION_HOST),
        "__KEY__": json.dumps(OPT_OUT_KEY),
        "__REGIONS__": json.dumps(list(CONSENT_DENIED_REGIONS)),
        "__LABELS__": json.dumps(labels, ensure_ascii=False, sort_keys=True),
    }
    script = LOADER
    for token, value in values.items():
        script = script.replace(token, value)
    return script


def _escape(value: str) -> str:
    return value.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")


def _slot() -> str:
    return '<span class="analytics-opt-out" hidden></span>'


def _note(locale: Locale, prefix: str) -> str:
    return (
        f'<p class="analytics-note" lang="{locale}" {MARKER}>'
        f"{_escape(text(locale, 'analytics_note'))} "
        f'<a href="{prefix}index.html#privacy-{locale}">'
        f"{_escape(text(locale, 'analytics_link'))}</a>{_slot()}</p>\n"
    )


def _disclosure(locale: Locale) -> str:
    return (
        f'<h3 id="privacy-{locale}" {MARKER}>'
        f"{_escape(text(locale, 'analytics_link'))}</h3>\n"
        f"<p>{_escape(text(locale, 'privacy_body'))}</p>\n"
        f"<p>{_escape(text(locale, 'privacy_cookies'))}</p>\n"
        f"<p>{_escape(text(locale, 'privacy_opt_out'))}{_slot()}</p>\n"
    )


def _replace_once(source: str, anchor: str, replacement: str, where: str) -> str:
    count = source.count(anchor)
    if count != 1:
        raise ValueError(f"{where}: expected one {anchor!r}, found {count}")
    return source.replace(anchor, replacement)


def integrity(script: str) -> str:
    """The Subresource Integrity value for one loader text."""
    digest = hashlib.sha384(script.encode("utf-8")).digest()
    return "sha384-" + base64.b64encode(digest).decode("ascii")


def add_to_page(source: str, relative: Path, pinned: str) -> str:
    """One page with GA added. ``relative`` is its path under the site root.

    ``pinned`` is the loader's :func:`integrity` value. Refuses (raises) a page
    that already carries the addition or whose shape it does not recognize,
    rather than guessing where a note belongs.
    """
    where = str(relative)
    if MARKER in source:
        raise ValueError(f"{where}: already carries Google Analytics")
    prefix = "../" * (len(relative.parts) - 1)
    script = (
        f'<script src="{prefix}{SCRIPT_NAME}" defer integrity="{pinned}" '
        f"{MARKER}></script>\n"
    )
    page = _replace_once(source, _HEAD_CLOSE, script + _HEAD_CLOSE, where)
    if relative.as_posix() == "index.html":
        for section_locale in LOCALES:
            opening = f'<section lang="{section_locale}"'
            start = page.find(opening)
            end = page.find(_LANDING_SECTION_END, start)
            if start < 0 or end < 0 or page.count(opening) != 1:
                raise ValueError(f"{where}: no single {opening!r} section")
            page = page[:end] + _disclosure(section_locale) + page[end:]
        return page
    match = _HTML_LANG.search(page)
    if match is None:
        raise ValueError(f"{where}: no <html lang> naming en or es")
    locale: Locale = "en" if match.group(1) == "en" else "es"
    note = _note(locale, prefix)
    if _FOOTER_END in page:
        return _replace_once(page, _FOOTER_END, note + _FOOTER_END, where)
    return _replace_once(page, _MAIN_END, note + _MAIN_END, where)


def add_to_tree(root: Path, raw_id: str = GA4_MEASUREMENT_ID) -> int:
    """Add GA to every page under ``root`` and write the loader. Returns pages changed.

    With no measurement ID configured this changes nothing, and says so.
    """
    measurement = measurement_id(raw_id)
    if measurement is None:
        return 0
    pages = sorted(root.rglob("*.html"))
    if not pages:
        raise ValueError(f"{root}: no pages to add Google Analytics to")
    script = loader(measurement)
    pinned = integrity(script)
    for path in pages:
        relative = path.relative_to(root)
        source = path.read_text(encoding="utf-8")
        path.write_text(add_to_page(source, relative, pinned), encoding="utf-8")
    (root / SCRIPT_NAME).write_bytes(script.encode("utf-8"))
    return len(pages)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Add Google Analytics 4 to a rendered site tree, in place."
    )
    parser.add_argument("root", type=Path, help="the rendered site directory")
    args = parser.parse_args(argv)
    changed = add_to_tree(args.root)
    if changed:
        print(f"analytics: Google Analytics 4 added to {changed} pages in {args.root}")
    else:
        print("analytics: no measurement ID configured; nothing added")
    return 0


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())
